### PR TITLE
Cache device class data

### PR DESCRIPTION
**Pull request by raman325**
bdraco shared this profile with me
image

We don't need to create the DeviceClass instance every time which will reduce the calls significantly. We also created `DeviceClassItem` on each lookup so I've stored it. If anything changes we will create a new DeviceClass instance on update so we shouldn't be storing any stale data in this PR

### CODE REVIEW

**Review by MartinHjelmare — APPROVED** [review]
Thanks!

**Comment by MartinHjelmare** [comment]
There's a merge conflict.

File: zwave_js_server/model/node/__init__.py [hunk 1]
```diff
@@ -417,30 +426,25 @@ def update(self, data: NodeDataType) -> None:
...
 
-        # Remove stale endpoints
-        for endpoint_idx in stale_endpoint_idxs:
-            self.endpoints.pop(endpoint_idx)
+        self._statistics = NodeStatistics(
+            self.client, self.data.get("statistics", DEFAULT_NODE_STATISTICS)
+        )
+        if last_seen := data.get("lastSeen"):
+            self._last_seen = datetime.fromisoformat(last_seen)
+        if not self._statistics.last_seen:
```
Comment (MartinHjelmare) [inline zwave_js_server/model/node/__init__.py:443]: Does this mean that last seen in the statistics can be different than the last seen of the latest data?

File: zwave_js_server/model/endpoint.py [hunk 2]
```diff
@@ -57,6 +57,7 @@ def __init__(
         self.node = node
         self.data: EndpointDataType = data
         self.values: dict[str, ConfigurationValue | Value] = {}
+        self._device_class: None | DeviceClass = None
```
Comment (MartinHjelmare) [inline zwave_js_server/model/endpoint.py:60]:
```suggestion
        self._device_class: DeviceClass | None = None
```

File: zwave_js_server/model/node/__init__.py [hunk 3]
```diff
@@ -115,6 +115,7 @@ def __init__(self, client: Client, data: NodeDataType) -> None:
             client, data.get("statistics", DEFAULT_NODE_STATISTICS)
         )
         self._firmware_update_progress: NodeFirmwareUpdateProgress | None = None
+        self._device_class: None | DeviceClass = None
```
Comment (MartinHjelmare) [inline zwave_js_server/model/node/__init__.py:118]:
```suggestion
        self._device_class: DeviceClass | None = None
```

**Review by MartinHjelmare — COMMENTED** [review]
Can we test the uncovered lines?

File: zwave_js_server/model/node/__init__.py [hunk 4]
```diff
@@ -371,26 +371,35 @@ def default_transition_duration(self) -> int | float | None:
...
-        if last_seen := data.get("lastSeen"):
-            self._last_seen = datetime.fromisoformat(last_seen)
-        if not self._statistics.last_seen:
-            self._statistics.last_seen = self.last_seen
+        # Remove stale endpoints
+        for endpoint_idx in stale_endpoint_idxs:
+            self.endpoints.pop(endpoint_idx)
+
+        # Add new endpoints or update existing ones
+        for endpoint_idx in new_endpoint_idxs - stale_endpoint_idxs:
```
[the remark this answers is not in the document]
Comment (raman325) [reply zwave_js_server/model/node/__init__.py:385]: good catch!

File: zwave_js_server/model/node/__init__.py (hunk 1 above)
Comment (raman325) [reply zwave_js_server/model/node/__init__.py:443]: no it means that the statistics last seen can be missing in which case we populate it to keep everything in sync

**Comment by raman325** [comment]
whoops sorry @MartinHjelmare didn't mean to merge this yet but hope this addressed all the concerns

**Comment by MartinHjelmare** [comment]
Looks good, thanks!

**Pull request merged**